pkg/kratos/service.go

```diff
@@ -163,6 +167,64 @@ func (s *Service) GetFlowError(ctx context.Context, id string) (*kClient.FlowErr
 	return flowError, resp.Cookies(), nil
 }
 
+func (s *Service) CheckAllowedProvider(ctx context.Context, loginFlow *kClient.LoginFlow, updateFlowBody *kClient.UpdateLoginFlowBody) (bool, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.Service.CheckAllowedProvider")
+	defer span.End()
+
+	provider := updateFlowBody.UpdateLoginFlowWithOidcMethod.Provider
+	clientName := loginFlow.Oauth2LoginRequest.Client.GetClientName()
+
+	allowed, err := s.authz.Check(ctx, fmt.Sprintf("app:%s", clientName), "allowed_access", fmt.Sprintf("provider:%s", provider))
+	if err != nil {
+		// We return false if an error occurs, this means that a fail on the authorization server will block our service
+		return false, err
+	}
+
+	return allowed, nil
+}
+
+func (s *Service) FilterFlowProviderList(ctx context.Context, flow *kClient.LoginFlow) (*kClient.LoginFlow, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.Service.FilterFlowProviderList")
+	defer span.End()
+
+	loginRequest := flow.Oauth2LoginRequest
+	// Extract providers
+	var providers []string
+	for _, node := range flow.Ui.Nodes {
+		switch node.Group {
+		case "oidc":
+			providers = append(providers, fmt.Sprintf("%v", node.Attributes.UiNodeInputAttributes.GetValue()))
+		}
+	}
+	clientName := loginRequest.Client.GetClientName()
+
+	// Filter providers
+	allowedProviders, err := s.authz.FilterObjects(ctx, fmt.Sprintf("app:%s", clientName), "allowed_access", "provider", providers)
+	if err != nil {
+		return nil, err
+	}
+
+	// This should happen only if the app has some allowed providers configured, but they don't match the ones on kratos
+	if len(allowedProviders) == 0 {
+		flow.Ui.Nodes = nil
+		return flow, ErrNoProvidersAllowed
+	}
+
+	// Filter UI nodes
+	var nodes []kClient.UiNode
+	for _, node := range flow.Ui.Nodes {
```

Comment:
the double loop feels redundant, but to remove it we d need to get the FilterObjects to accept nodes and return nodes which right now seems overkill

still, sounds redundant

Reply:
Yes I thought about that, but even in that case you would push the loops in the `FilterObjects`, I don't think we can avoid it.